### PR TITLE
Refactor window closing and app exit handling

### PR DESCRIPTION
**Pull request by obulat**
Signed-off-by: Olga <obulat@gmail.com>

When #741 added asyncio handling in winforms backend, it accidentally removed lines that were used to handle app exit event (added in #522) and to show a secondary window (added in #432). This PR adds them back. I tested adding an `on_exit` handler, but didn't test adding secondary window.

## PR Checklist:

- [x] All new features have been tested
- [ ] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

**Comment by obulat** [comment]
> We could probably add this to the existing dialogs demo - in addition to all the other "open dialog" buttons, add a "show new window" button that pops up a new window. Include a counter in the title of the window (so you get "New window 1" then "New window 2"); and hook in some on close logic on the window so that when it is closed, it updates a label on the main window that says which window was closed.

I realized that this example needs additional functionality: we had `on_exit` for main_window closing, but no handler for the event of closing a secondary window. I've added on_close handler to core and winforms backend.
windows

File: examples/dialogs/dialogs/app.py [hunk 1]
```diff
@@ -90,12 +90,53 @@ def action_save_file_dialog(self, widget):
         except ValueError:
             self.label.text = "Save file dialog was canceled"
 
+    def action_open_secondary_window(self, widget):
+        # Problems:
+        # - Setting secondary window's `app` manually
```
Comment (freakboy3742) [inline examples/dialogs/dialogs/app.py:95]: This is a good point - how do we tell the window about the app?

The MainWindow gets assigned to the app when you set `self.main_window = ...`. I wonder if we should do a similar thing for other windows, and require that they are explicitly added to the app:

    self.windows += toga.Window(...)

or:

    self.windows.add(toga.Window(...))


and raise an error if you attempt to call `show()` on a window that doesn't have an app assigned. 

In this world, `TogaApp.windows` would be a set-like object that has add/remove methods (including support for the + and += operators); adding a window to `TogaApp.windows` would assign the window to the app.

Does that make sense as an API?

File: examples/dialogs/dialogs/app.py [hunk 2]
```diff
@@ -90,12 +90,53 @@ def action_save_file_dialog(self, widget):
...
+        # - Secondary window's `icon` is not set
```
Comment (freakboy3742) [inline examples/dialogs/dialogs/app.py:96]: I think this is a windows-only requirement; would it make sense to use the App's icon? Or do Windows apps have different icons for each window?

File: examples/dialogs/dialogs/app.py [hunk 3]
```diff
@@ -90,12 +90,53 @@ def action_save_file_dialog(self, widget):
...
+            cancel_exit()
+            self.label.text = 'Exit canceled'
+
+    def set_window_label_text(self):
+        self.window_label.text = '{} secondary windows open'.format(self.window_count - 1)
+
     def startup(self):
         # Set up main window
         self.main_window = toga.MainWindow(title=self.name)
+        self.window_count = 1
```
Comment (freakboy3742) [inline examples/dialogs/dialogs/app.py:131]: It might be better to keep this as a "secondary window count" - that way we can avoid doing the "and subtract one" every time we actually use it.

File: examples/dialogs/dialogs/app.py [hunk 4]
```diff
@@ -90,12 +90,53 @@ def action_save_file_dialog(self, widget):
...
+                flex=1,
+                direction=COLUMN,
+                padding=10
+            )
+        )
+        window.app = self
+        window.on_close = close_handler
+        window.show()
+
+    def exit_handler(self, app, cancel_exit):
```
Comment (freakboy3742) [inline examples/dialogs/dialogs/app.py:118]: This is a good idea, but we need to differentiate two things here: an event handler that is triggered when an app *is* exiting, and an event handler that is triggered to determine if an app *can* exit. Some platforms do this with two different handlers (macOS uses willClose and didClose); others use a boolean return value from the handler (true means the app can exit). 

Passing in a callable definitely works in the windows case, but I'm not sure it's the best API overall.

File: examples/dialogs/dialogs/app.py (hunk 3 above)
Comment (obulat) [reply examples/dialogs/dialogs/app.py:131]: Great point!

File: examples/dialogs/dialogs/app.py (hunk 2 above)
Comment (obulat) [reply examples/dialogs/dialogs/app.py:96]: I'm not sure what the recommendations on Form icons are. I would like to look at examples and see what common usage is, but I can't think of apps that have secondary windows. There are also different type of secondary windows available, including ones that are set to not have an icon:

> The Form class can be used to create standard, tool, borderless, and floating windows. You can also use the Form class to create modal windows such as a dialog box. 

I could just set the App's icon as each Form's icon for now.

**Comment by obulat** [comment]
> This demo is really good; I've flagged a couple of minor things, but for the most part it's really good.

Thank you!

>  I'm almost inclined to suggest that we don't need on_close _and_ on_exit - we only need on_close, with on_close on the main window being application close. I'd also be inclined to use the return value of on_close as the trigger for whether the window/app should be allowed to close.

Do you mean that we don't need on_close and on_exit on the interface level, or the user-facing API? I believe that the user-facing API needs both, and on the interface level we can handle both with a single on_close handler like this:
```
def on_close():
    if window is main window:
        self._impl.on_exit()
    else:
        self._impl.on_close()
``` 
Is this what you meant?

**Comment by freakboy3742** [comment]
I originally thought that we may not need an public facing `on_exit`, but thinking about it some more, I think it's worth having a separate `on_exit` handler - if only because Document based apps (like Podium) don't have a main window, but obviously they still exit.

Part of the cleanup here needs to be clarifying the difference between the "internal" implementation on_close and the "public" interface on_close. The fact that we're having confusion over which on_close method we need is an indication of a deeper problem :-) The "gtk_*"/"winforms_*" convention for native signals might be a pattern to emulate here; if we pick a prefix for "internal" signals, we can avoid colliding in future.

File: examples/dialogs/dialogs/app.py (hunk 1 above)
Comment (obulat) [reply examples/dialogs/dialogs/app.py]: I've added a possible implementation of self.windows, could you take a look?
The example doesn't count windows correctly because the handling of window on_close is incorrect at the moment.

File: examples/dialogs/dialogs/app.py (hunk 2 above)
Comment (obulat) [reply examples/dialogs/dialogs/app.py]: I added the app's icon to secondary windows in show().

File: examples/dialogs/dialogs/app.py (hunk 4 above)
Comment (obulat) [reply examples/dialogs/dialogs/app.py]: > I originally thought that we may not need an public facing `on_exit`, but thinking about it some more, I think it's worth having a separate `on_exit` handler - if only because Document based apps (like Podium) don't have a main window, but obviously they still exit.

When I tried to implement Document-based app (Podium) on Windows, I used a main Form that has IsMdiContainer (is Multiple Document Interface Container) set to true for the container window, and set it as the main window; and secondary windows (Forms) that have MdiParent property set to the container Form. What's the best way of handling this difference between Mac and Windows re: main window presence/ absence?

Comment (freakboy3742) [reply examples/dialogs/dialogs/app.py]: So, I'm not 100% sure what the "right" approach is on Windows. 

On macOS, the App runs exactly once; it always has a menu at the application level; and if you haven't got a document open, a file dialog opens prompting you to open a document. Each Document that is open has a primary window.

On Linux, each "document" is a separate instance of the running app. Each document has a main window with a menu.

I haven't used Windows for a while, so I'm not 100% sure what users would expect - whether it's closer to the Apple model (but with a menu on each document window), or Linux, somewhere between the two, or something completely different. I remember MDI apps from the days of Windows 3.1 - but that was a "main" window that contained *internal* sub-windows - but that style of app hasn't been common for a long time, AFAIK. If Winforms concept of MDI is different, it might be the right approach; but it would be helpful to see what that looks like to compare with other Windows apps.

The best suggestion I've got is to step away from Toga for a moment, and try and build an app directly with Python.NET that has the right behavior. The app doesn't need to actually *do* anything - putting up a window with a label that says "I am document N". That way we can make sure we have the basic ideas right before we worry about integrating with Toga.

Comment (freakboy3742) [reply examples/dialogs/dialogs/app.py]: I got interested, and dug into this a little more. If Notepad is any indication, the Linux approach might be the right one. If you open to two documents, there are two instances of notepad.exe. So, that should make things a lot easier - we can keep the same "main" window infrastructure, and wrap it with some file handling. The Linux implementation should also be a fairly good point of reference.

Comment (obulat) [reply examples/dialogs/dialogs/app.py]: I think you are right that Windows is closer to Linux here. One point I would add is usually, when you have a document open in an app, and want to open a new one, Windows doesn't start a new app instance, it just replaces the document in the main window of the previous app. Maybe, @brettcannon could answer with more certainty?..

Comment (brettcannon) [reply examples/dialogs/dialogs/app.py]: I am most definitely not a Windows expert so I can't really answer what the UX should be here.

Comment (obulat) [reply examples/dialogs/dialogs/app.py]: Oh, sorry about this misunderstanding on my part.

File: src/core/toga/window.py [hunk 5]
```diff
@@ -181,9 +186,19 @@ def full_screen(self, is_full_screen):
     def close(self):
         self._impl.close()
 
-    def on_close(self):
+    def toga_on_close(self):
+        self.app.windows -= self
         self._impl.on_close()
```
Comment (freakboy3742) [inline src/core/toga/window.py:191]: I can see what you're doing here; however, the old "on_close()" method was probably a mistake. By having it on the interface layer, we're effectively saying "this is a method you *could* invoke" - which definitely isn't true. 

Since all we're ever going to be doing here is:
1. removing the window from the `self.windows` list, and
2. triggering the *user's* `on_close` handler
we may be better off leaving this logic 100% on the implementation side. Yes, this does mean we're going to be duplicating the "remove window, invoke hander" logic on every backend... but then, that's already true for every other event handler anyway, and it allows us to customise exactly how and when the events are triggered (in the unlikely event that ordering ever matters).

Comment (saroad2) [reply src/core/toga/window.py:191]: Why not simply making it private with the "__" prefix? isn't it enough in order to indicate that this is not a method that should be called?

Comment (obulat) [reply src/core/toga/window.py]: I transferred the `on_close` logic to implementation side, basically copying `button`'s `on_press` handler behavior... I've tested it on Windows, and ran the tests (updated). I also added the same code to gtk and cocoa, but I cannot test them, so I don't know if they work.

**Comment by obulat** [comment]
I've pushed some additional changes just to make the tests pass, but this way there are too many 'close' related functions everywhere.

I think I've finally understood what you meant when you said we don't need a separate `on_exit` handler. We can set separate `on_close` handlers for each window (main and secondary) separately, and the `on_close` for main window would handle the `exit` case. 

And yes, for Winforms this makes sense, and we don't need a separate handler. But we need it for document apps. So, what would be the best API for users? 

I had an idea that `on_exit` is for things you need to finish up before closing the app: saving some data, and maybe prompting the user if they really want to exit. For these actions, the word 'exit' seems more appropriate. And this distinction would be really clear between actions that need to happen when the app with all of its windows is closing, and when simply one window closes (when the window's `on_close` handler would be called).

**Comment by saroad2** [comment]
Fixed merge conflicts by merging from the master branch, and added that when `App.exit` is called, `App.on_exit` should be called if it exists.

Hope that's OK. If not, feel free to revert those commits.

**Comment by saroad2** [comment]
Merged from master to solve merging conflicts.

File: src/core/toga/app.py [hunk 6]
```diff
@@ -22,6 +23,49 @@
...
+    notations. Adding a window to app automatically sets `window.app` property to the app.
+    """
+    def __init__(self, app, iterable=None):
+        self.app = app
+        self.elements = set() if iterable is None else set(iterable)
+
+    def add(self, window: Window) -> None:
+        if not isinstance(window, Window):
+            raise TypeError("Toga app windows can only add objects of toga.Window type")
+        # Silently not add if duplicate
```
Comment (obulat) [inline src/core/toga/app.py:39]: What is the best action when we call add/discard for a window that is not associated with the app?

**Comment by obulat** [comment]
I've added cocoa support and testing. `on_close` for window and `on_exit` for app seem to work now.
One question I still have is what is the best action for adding/removing a window that hasn't been associated with the app?

```python
app = toga.App()
window = toga.Window()
app.windows -= window # Action???
```

Should we:
1. just silently not do anything,
2.  raise an exception,
3. notify the user without raising an exception ?

File: src/core/toga/app.py (hunk 6 above)
Comment (freakboy3742) [reply src/core/toga/app.py:39]: Adding a window that isn't associated with the app would be 'normal' behavior. Adding a window if it's already associated with a *different* app should be an exception - same as adding a widget when it's already associated with a different app - but this should be almost impossible to actually do in practice.

I'd argue *removing* a window that isn't associated with this app should be an exception, since it indicates the app has got into a weird state (probably because of something the developer has done tracking their windows).

File: src/core/toga/app.py [hunk 7]
```diff
@@ -226,6 +273,11 @@ def __init__(
         self._startup_method = startup
 
         self._main_window = None
+        # In this world, TogaApp.windows would be a set-like object
+        # that has add/remove methods (including support for
+        # the + and += operators); adding a window to TogaApp.windows
+        # would assign the window to the app.
+        self.windows = Windows(self, windows)
```
Comment (freakboy3742) [inline src/core/toga/app.py:280]: To be clear - `self.windows` makes sense as a variable name; it's the data type (`Windows`) we should change.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
The basic mechanics of adding/removing windows look good to me; I've addressed your question about adding/removing a window that isn't associated with an app inline.

However, I saw some inconsistent behavior in the `dialogs` example.

On Windows, hitting the close button on a window causes the "are you sure" dialog to be shown, as expected. If you say "Cancel", the close is aborted; however, if you say "OK", the main window closes, but the confirmation dialog is shown a second time. The behavior otherwise looks good to me. I suspect that it's probably getting tied in knots between behaviours that need to execute when the user requests a window to close, and behavior that executes when the main window closes in response to quitting the app.

On macOS, no confirmation dialog is shown at all if you click on the close button for the main window; however, the app is put into a weird limbo state where it doesn't have a main window, but it hasn't been closed either. The main menu "Quit" option still works, though; and that option shows the confirmation dialog.

The macOS behavior is what I'd expect from a *document* based app, but the dialogs demo is a Main window-based app; closing the main window should be the same as "quit".

File: examples/dialogs/dialogs/app.py [hunk 8]
```diff
@@ -104,12 +104,60 @@ def action_save_file_dialog(self, widget):
...
+        else:
+            self.label.text = 'Exit canceled'
+            return 'cancel'
+
+    def set_window_label_text(self):
+        self.window_label.text = f"{len(self.windows) - 1} secondary window(s) open"
+
     def startup(self):
         # Set up main window
         self.main_window = toga.MainWindow(title=self.name)
```
Comment (obulat) [inline examples/dialogs/dialogs/app.py:153]: User can also add the `on_close` handler to the main_window here.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
This is starting to look really good. I've flagged a few things inline; I'll push an update shortly that addresses most of them. 

I still need to test GTK; but in the meantime, there's two issues pending that I can see. 

1) There's a weird edge case in Cocoa handling. `MainWindow.toga_on_close` invokes `exit`, rather than running the on_exit handler. As a result, it's possible to *approve* the main window to close, but *deny* the app from exiting - which results in a running app with no main window. 

2) On winforms, the close action doesn't seem to fully honour a "don't close" from the MainWindow. If you say you don't want to close the main window, there's no point to asking if you want the app to exit.

This does make me wonder whether we should perhaps *not* have a user-space on_close handler for MainWindow - since on_close on the main window *is* an exit, it perhaps doesn't make sense to have both.

**Review by obulat — COMMENTED** [review]
> This does make me wonder whether we should perhaps _not_ have a user-space on_close handler for MainWindow - since on_close on the main window _is_ an exit, it perhaps doesn't make sense to have both.

I've been thinking about it the use-cases for having separate `on_close` for the MainWindow and `on_exit` for the App in a non-Document-app, and I don't see any. I think only having `on_close` for secondary windows, and `on_exit` for the MainWindow closing is the best option, I will add that.

**Comment by freakboy3742** [comment]
One additional suggestion around removing `on_close` for MainWindow - since MainWindow is a subclass of Window, there will still be an API that would allow a user to *set* `on_close`. Ideally, we'd either disable that capability for `on_close` - or, alternatively, raise an error if someone tries to set an `on_close` handler for a main window (or, at the very least, have the `toga_on_close` platform handler look for an `on_close` and raise a warning that it's going to be ignored)

**Comment by obulat** [comment]
I removed the on_close from the main window, leaving only the app's on_exit. Trying to set an on_close on the main window raises an AttributeError (?) now.
There was a problem in Winforms where when the window closing is canceled, the window is closed anyways. I fixed that, and also fixed the example app showing the wrong number of open secondary windows in the label.

**Review by freakboy3742 — APPROVED** [review]
This looks really good - thanks for all the hard work!

I've made a couple of small tweaks:
* I've added some docstrings to explain why how the example app works
* I've consolidated the handling of `on_exit()` into the interface layer - after all, every app should be calling `on_exit` in the same way
* I've tweaked the naming of the `toga_on_close` handler. For every other event, we've used the platform prefix (`gtk_`, `winforms_` etc) to identify internal system native events; in this patch, all the event handlers are internal, so it makes sense (to me, anyway!) to preserve that naming. That also helps make it clear when you're handling a "toga" event or a "native" event.
* I've tweaked the logic around the GTK backend - it wasn't doing the `on_exit` check on app close.

Thanks for all your persistence on this - it's been a long time in the making, and we wouldn't have got here without all your effects!
